### PR TITLE
MPI collective

### PR DESCRIPTION
**Pull request by rubenhorn**
Bitwise reproducibility of results achieved with broadcast, but not with reduce operation. (Might not even be possible due to lack of associativity.)  
I am not sure how necessary that is for MaMiCo. Thoughts?

See #8 for more details.

### CODE REVIEW

**Comment by rubenhorn** [comment]
When testing on HSUper: Stuck after `Finish CouetteScenario::initSolvers()` (in `runOneCouplingCycle`)

**Comment by rubenhorn** [comment]
> Benchmarking with different scenarios (multiple nodes, non-trivial decomposition) gets stuck after `Finish CouetteScenario::initSolvers()` (in `runOneCouplingCycle`)

Blocking on `MPI_Comm_create_group` in `SendReceiveBuffer::allocateBcastRequests` on HSUper with IntelMPI 2021.6.
(Potential issue: https://stackoverflow.com/questions/63580093/is-there-a-non-blocking-version-of-mpi-comm-create-group)

Results from running on Minicluster AMD directly using `mpirun -H amd00:64,amd01:64,amd02:64,amd03:64` ($4^3$ processes) with ls1 MD60/100 and 4 MD instances for 750 coupling cycles:

Collective:
```
Finished all coupling cycles after 1021.98 s
Time percentages Micro, Macro, Filter, Other: 
95.9112, 0.391634,  3.07246e-05, 3.69711
```

Reference:
```
Finished all coupling cycles after 994 s
Time percentages Micro, Macro, Filter, Other: 
98.0143, 0.405366,  1.28773e-05, 1.58029
```

:warning: Scenario likely too small for collective communication (:point_right:  overhead dominates)

**Comment by Thinkpiet** [comment]
Could you please

- Stay on amd00:64,amd01:64,amd02:64,amd03:64 (ignore IntelMPI on HSUper for now)
- Go down to ls1 MD30/50 (otherwise we just waste compute time for things we are not interested in here)
- Go down to 1x1x1 spatial subdomains / processes i.e. sequential in space; keep this constant
- Run a series of experiments for n \in N={1,4,16,64,100,128,200,256} MD instances, scale the number of MPI processes accordingly
- Collect the values of X of the "Finished all coupling cycles after X s" for collective X_c and for reference X_r
- Post the resulting plot of X_c and X_r over n ?

Expected results: 
- X_c(1) might be a bit slower / higher than X_r(1), like 3% or so
- But X_r(n) will rise linearly with n
- While X_c(n) should be constant, X_c(1)==X_c(256), hopefully

**Comment by rubenhorn** [comment]
Preliminary findings from the WindHPC cluster because I'm still having issues running it on Minicluster through Slurm:
image
It seems that the collective implementation indeed requires less time, but sample and the scenario are too small (it does not make a dent in the job time) to be certain and to quantify it.

**Comment by rubenhorn** [comment]
Stale branch [mpi-collective ](https://github.com/HSU-HPC/MaMiCo/tree/mpi-collective) should be deleted as well.

File: coupling/sendrecv/SendReceiveBuffer.cpph [hunk 1]
```diff
@@ -35,8 +61,616 @@ void coupling::sendrecv::SendReceiveBuffer<MacroscopicCell, dim>::deleteBuffer(s
...
+    }
+  } // rank
+}
+
+template <class MacroscopicCell, unsigned int dim>
+void coupling::sendrecv::SendReceiveBuffer<MacroscopicCell, dim>::triggerBcasts(const unsigned int thisRank) {
+
+#if (COUPLING_MD_PARALLEL == COUPLING_MD_YES)
+  int counter = 0;
+  // if(thisRank == 23) std::cout << thisRank << " sendrecvBuffer size " << _sendRecvBuffer.size() << std::endl;
```
Comment (LouieVoit) [inline coupling/sendrecv/SendReceiveBuffer.cpph:239]: Maybe clean up the comments

File: coupling/IndexConversion.cpph [hunk 2]
```diff
@@ -111,6 +111,14 @@ template <unsigned int dim> tarch::la::Vector<dim, unsigned int> coupling::Index
   return _parallelTopology->getProcessCoordinates(rank);
 }
 
+template <unsigned int dim> tarch::la::Vector<dim, unsigned int> coupling::IndexConversion<dim>::getThisProcessCoordinates(unsigned int rank) const {
+  tarch::la::Vector<dim, unsigned int> processCoordinates = getProcessCoordinates(rank);
+  for (unsigned int d = 0; d < dim; ++d) {
+    processCoordinates[d] = processCoordinates[d] % _numberProcesses[d];
```
[the remark this answers is not in the document]
Comment (Thinkpiet) [reply coupling/IndexConversion.cpph:117]: Only if sqrt(d) < 0

**Pull request merged**